webui/src/lib/components/repository/ObjectsDiff.jsx

```diff
@@ -0,0 +1,167 @@
+import React from "react";
+import {useAPI} from "../../hooks/api";
+import {objects} from "../../api";
+import ReactDiffViewer, {DiffMethod} from "react-diff-viewer";
+import {Error, Loading} from "../controls";
+
+export const ObjectsDiff = ({diffType, repoId, leftRef, rightRef, path}) => {
+    switch (diffType) {
+        case 'changed':
+            return <ChangedDiff repoId={repoId} leftRef={leftRef} rightRef={rightRef} path={path}/>;
+        case 'added':
+            return <AddedDiff repoId={repoId} rightRef={rightRef} path={path}/>;
+        case 'removed':
+            return <RemovedDiff repoId={repoId} leftRef={leftRef} path={path}/>;
+        case 'conflict':
+            return <ConflictDiff path={path}/>;
+        default:
+            return <Error error={"Unsupported diff type " + diffType}/>;
+    }
+}
+
+// github shows 20KB diff according to https://github.blog/2016-12-06-how-we-made-diff-pages-3x-faster/
+const maxDiffSizeBytes = 20000;
+const supportedReadableFormats = ["txt", "csv", "tsv"];
+
+function readableObject(path) {
+    for (let i = 0; i < supportedReadableFormats.length; i++) {
+        if (path.endsWith(supportedReadableFormats[i])) {
+            return true;
+        }
+    }
+    return false;
+}
+
+const ConflictDiff = ({path}) => {
+    return "Conflict in " + path + " resolve the conflict manually to review objects diff.";
+}
+
+const ChangedDiff =  ({repoId, leftRef, rightRef, path}) => {
+    const readable = readableObject(path);
+    const getFn = readable ? objects.getWithSizeConstraint : objects.getStat;
+    const left = useAPI(async () => getFn(repoId, leftRef, path, maxDiffSizeBytes),
+        [repoId, leftRef, path]);
+    const right = useAPI(async () => getFn(repoId, rightRef, path, maxDiffSizeBytes),
+        [repoId, rightRef, path]);
+
+    if (left.loading || right.loading) return <Loading/>;
+    if (!!right.error) return <Error error={right.error}/>;
+    if (!!left.error) return <Error error={left.error}/>;
+
+    return readable
+        ? <ContentDiff left={left.response[0]} right={right.response[0]} leftSize={(left.response[1]).get("Content-length")}
+                       rightSize={(right.response[1]).get("Content-length")} diffType={"changed"}/>
+        : <StatDiff left={left.response} right={right.response} diffType={"changed"}/>;
+}
+
+const AddedDiff =  ({repoId, rightRef, path}) => {
+    let readable = readableObject(path);
```

Comment:
A lot of code is duplicated between those components. 
In the end all of them (except conflict) return a ContentDiff or a StatDiff. Consider merging these into one component.